### PR TITLE
Spring Security

### PR DESCRIPTION
**Pull request by Eidenn005**
Added Spring security
Added Roles and needed classes
Change User class according to security

### CODE REVIEW

**Review by DmytroHadiuchko — APPROVED** [review]
Good job!

File: src/main/java/com/example/onlinebookstore/security/CustomUserDetailsService.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException("Can`t find user by email"));
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/security/CustomUserDetailsService.java]:
```suggestion
                .orElseThrow(() -> new UsernameNotFoundException("Can`t find user by email: + EMAL"));
```

File: pom.xml [hunk 2]
```diff
@@ -77,6 +77,12 @@
 			<groupId>org.springframework.boot</groupId>
 			<artifactId>spring-boot-starter-validation</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-security</artifactId>
+			<version>3.2.5</version>
+		</dependency>
+
```
Comment (andrii-hoienko) [inline pom.xml]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 3]
```diff
@@ -35,11 +36,13 @@ public BookDto getBookById(@PathVariable Long id) {
 
     @ResponseStatus(HttpStatus.CREATED)
     @PostMapping
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java:46]: Why there is no annotation PreAuthorize on endpoints that should be accessible to user

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left some minor comments

File: src/main/java/com/example/onlinebookstore/controller/BookController.java (hunk 3 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/controller/BookController.java:46]: Added

File: pom.xml (hunk 2 above)
Comment (Eidenn005) [reply pom.xml]: done

File: src/main/java/com/example/onlinebookstore/security/CustomUserDetailsService.java (hunk 1 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/security/CustomUserDetailsService.java]: done

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/api/auth/**")
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/config/SecurityConfig.java]: also add swagger and `/error`

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 5]
```diff
@@ -29,25 +30,29 @@ public class BookController {
 
     @GetMapping
     @Operation(summary = "Get all books", description = "Get list of all available books")
+    @PreAuthorize("hasAnyRole('ROLE_USER','ROLE_ADMIN')")
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java]: we need to add `ROLE_` when we use `hasAuthorithy()` but here we already specified that we are checking exactly role.
Same for other similar cases
```suggestion
    @PreAuthorize("hasAnyRole('USER','ADMIN')")
```

File: src/main/java/com/example/onlinebookstore/repository/RoleRepository.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.example.onlinebookstore.repository;
+
+import com.example.onlinebookstore.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Role findByRole(Role.RoleName role);
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/repository/RoleRepository.java]:
```suggestion
    Optional<Role> findByRole(Role.RoleName role);
```

File: src/main/java/com/example/onlinebookstore/model/Role.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "roles")
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/model/Role.java:18]: I do not see sql cripts for creation of this and other tables that are part of this task, It also mentions that we should insert some data into those tables, at least `ADMIN` user

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.security.web.SecurityFilterChain;
+
+@EnableMethodSecurity
+@Configuration
+public class SecurityConfig {
+    private UserDetailsService userDetailsService;
+
+    public SecurityConfig(UserDetailsService userDetailsService) {
+        this.userDetailsService = userDetailsService;
+    }
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/config/SecurityConfig.java:22]: we can use lombok instead, but up to you

**Review by karma-o — CHANGES_REQUESTED** [review]
Check task description again for direct sql data insertion

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java (hunk 4 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/config/SecurityConfig.java]: done

File: src/main/java/com/example/onlinebookstore/controller/BookController.java (hunk 5 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/controller/BookController.java]: done

File: src/main/java/com/example/onlinebookstore/repository/RoleRepository.java (hunk 6 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/repository/RoleRepository.java]: Done

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java (hunk 8 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/config/SecurityConfig.java:22]: Did it in next task.

File: src/main/java/com/example/onlinebookstore/model/Role.java (hunk 7 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/model/Role.java:18]: done

**Review by andrii-hoienko — APPROVED** [review]
Good job!

**Pull request merged**